src/integrationtests/java/com/aws/greengrass/integrationtests/ShadowManagerTest.java

```diff
@@ -247,4 +263,58 @@ void GIVEN_shadow_manager_WHEN_strategy_config_resets_THEN_respond_to_config_upd
         kernel.getContext().waitForPublishQueueToClear();
         assertThat(syncHandler.getOverallSyncStrategy(), instanceOf(PeriodicSyncStrategy.class));
     }
+
+    @Test
+    void GIVEN_shadow_manager_running_AND_crypto_service_loading_WHEN_sync_requests_with_dataplane_client_THEN_retry_requests_with_good_client_after_crypto_service_loads(ExtensionContext context)
+            throws InterruptedException, TLSAuthException, IoTDataPlaneClientCreationException {
+        ignoreExceptionOfType(context, TLSAuthException.class);
+        ignoreExceptionOfType(context, RetryableException.class);
+
+        IotDataPlaneClientFactory factory = kernel.getContext().get(IotDataPlaneClientFactory.class);
+        IotDataPlaneClientWrapper wrapper = spy(new FakeIotDataPlaneClientWrapper(factory));
+        kernel.getContext().put(IotDataPlaneClientWrapper.class, wrapper);
+
+        // Shadow manager starts syncing 1 shadow in the realtime as per the recipe and getThingShadow
+        // cloud api is called as part of syncing.
+        startNucleusWithConfig(NucleusLaunchUtilsConfig.builder()
+                .configFile("sync.yaml")
+                .mqttConnected(true)
+                .mockCloud(false)
+                .build());
+        BaseSyncStrategy syncStrategy = kernel.getContext().get(RealTimeSyncStrategy.class);
+        assertThat("syncing has started", syncStrategy::isSyncing, eventuallyEval(is(true)));
+        verify(wrapper, timeout(5000).atLeast(1)).getThingShadow("Thing1", "");
+
+        // This test ensures that the request is retried again as the we're not mocking cloud and data plane client
+        // creation fails. If the data plane client creation succeeds, the request is processed once again.
+        CountDownLatch cdl = new CountDownLatch(1);
+        SecurityService securityService = mock(SecurityService.class);
+        when(securityService.getDeviceIdentityKeyManagers()).thenAnswer((invocation) -> {
+            cdl.countDown();
+            return new KeyManager[0];
+        });
+        kernel.getContext().put(SecurityService.class, securityService);
```

Comment:
Maybe I'm missing something. Isn't there a race if you don't store this in the context before starting nucleus?

Reply:
Oh, ok, so you're essentially hot patching and then relying on an implementation detail (that the security service is retrieved from the context rather than cached).

Can we inject this mock up front and then update the response? I think that achieves the same but without relying on that implementation detail in Nucleus